File: 1290-get-decimal-value/Solution.java
//自己手寫的版本, 先知道整個鏈表的size才有辦法做運算
//另外一版, 不用知道鏈表的整個size也能求解, 用python寫
class ListNode{
    int val;
    ListNode next;
    ListNode prev;
    ListNode(int val,ListNode prev, ListNode next){
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}
class LinkedList{
    ListNode head;
    ListNode tail;
    void AddAtTail(ListNode newNode){
        if (head == null){
            head = newNode;
            tail = newNode;
        }
        newNode.prev = tail;
        tail.next = newNode;
        tail = newNode;
    }
}

class Solution1290 {
    public int getDecimalValue(ListNode head) {
        ListNode dummy = new ListNode(0, null, head);
        ListNode cur = dummy;
        int size = -1, result = 0;

        while (cur.next != null){
            cur = cur.next;
            System.out.println(cur.val);
            size++;
        }
        System.out.println(size);

        cur = dummy;
        while (cur.next != null){
            cur = cur.next;
            if (cur.val == 1){
                for (int j = size; j > 0; j--){
                    cur.val *= 2;
                }
            }
            size--;
            System.out.println(cur.val);
            result += cur.val;
        }

        System.out.println(result);
        return result;
    }

    static void main(String[] args) {
        int[] nums = {1, 0, 0, 1};
        LinkedList mainList = new LinkedList();
        for (int i : nums){
            ListNode newNode = new ListNode(i, null, null);
            mainList.AddAtTail(newNode);
        }
        Solution1290 sol = new Solution1290();
        sol.getDecimalValue(mainList.head);
    }
}

File: 1290-get-decimal-value/solution.py
from typing import Optional
class ListNode:
    def __init__(self, val, next) -> None:
        self.val = val
        self.next = next
class LinkedList:
    def __init__(self) -> None:
        self.head = None
    def AddAtTail(self, newNode):
        if self.head is None:
            self.head = newNode
            return
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = newNode
class Solution1122:
    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        if head is None:
            return None
        cur = head
        ans = 0
        while cur is not None:
            ans = ans * 2 + cur.val
            print("ans = ", ans)
            cur = cur.next
            if cur is None:
                return ans
if __name__ == "__main__":
    nums = [1, 0, 1]
    sol = Solution1122()
    mainList = LinkedList()
    for i in nums:
        newNode = ListNode(i, None)
        mainList.AddAtTail(newNode)
    sol.getDecimalValue(mainList.head)